Orleans.Providers.MongoDB/Reminders/Store/MongoReminderCollection.cs

```
@@ -35,48 +37,66 @@ protected override string CollectionName()
-            try
+            List<string> indexesToRemove =
+            [
+                // ByHash definitions.
+                "ByHash",
+#pragma warning disable CS0618 // Type or member is obsolete
+                $"{collection.GetFieldName(r => r.IsDeleted)}_1"
+                + $"_{collection.GetFieldName(r => r.ServiceId)}_1"
+                + $"_{collection.GetFieldName(r => r.GrainHash)}_1"
+#pragma warning restore CS0618 // Type or member is obsolete
+            ];
+
+            if (cleanAllRedundantIndexes)
             {
-                collection.Indexes.CreateOne(
-                   new CreateIndexModel<MongoReminderDocument>(byNameDefinition,
-                        new CreateIndexOptions
-                        {
-                            Name = "ByName"
-                        }));
+                indexesToRemove.AddRange([
+                    // ByName definition
+                    "ByName",
+#pragma warning disable CS0618 // Type or member is obsolete
+                    $"{collection.GetFieldName(r => r.IsDeleted)}_1"
+                    + $"_{collection.GetFieldName(r => r.ServiceId)}_1"
+                    + $"_{collection.GetFieldName(r => r.GrainId)}_1"
+                    + $"_{collection.GetFieldName(r => r.ReminderName)}_1",
+#pragma warning restore CS0618 // Type or member is obsolete
+                ]);
             }
-            catch (MongoCommandException ex)
+
+            foreach (var indexToRemove in indexesToRemove)
             {
-                if (ex.CodeName == "IndexOptionsConflict")
+                // best effort drop
+                try
+                {
+                    collection.Indexes.DropOne(indexToRemove);
+                }
+                catch (MongoCommandException e)
                 {
-                    collection.Indexes.CreateOne(new CreateIndexModel<MongoReminderDocument>(byNameDefinition));
+                    if (!e.WasIndexNotFound()) throw;
```

Re-throwing here will completely fail reminder service initialisation and likely kill the silo. I think we're better off failing siltently, what's the worst that could happen if dopping the index fails for any reason...

---

Yes, it would. But it was already previous behavior. If these lines raised an exception (e.g. another index was created with a non-default name by an opinionated MongoDB admin), the silo would have died. Same with standard Mongo reasons - network not available, ect.

https://github.com/OrleansContrib/Orleans.Providers.MongoDB/blob/master/Orleans.Providers.MongoDB/Reminders/Store/MongoReminderCollection.cs#L54-L57

https://github.com/OrleansContrib/Orleans.Providers.MongoDB/blob/master/Orleans.Providers.MongoDB/Reminders/Store/MongoReminderCollection.cs#L77-L80

---

Agreed, but should this be the behavior going forward? or do we want err on the side of safety?

---

In a few prior discussions, we erred on the policy of following pre-existing Orleans providers behaviors. All the other providers also throw when not dealing with specific paths.

Consider these two drills into the stack trace.

https://github.com/dotnet/orleans/blob/5bfe38820f7f83118d2cca799b14f720bd84bcba/src/Azure/Orleans.Reminders.AzureStorage/Storage/AzureBasedReminderTable.cs#L39-L70

https://github.com/dotnet/orleans/blob/7a05d489b429fccf3e804ac4ef9680a87f91a10e/src/Azure/Shared/Storage/AzureTableDataManager.cs#L75-L104

---

Yes, but in the code you've referenced, it's failing to initialise the reminders table which is detrimental to its on-going operations. In this case, if the only failure is in dropping old un-used indexes, then I don't think it should affect the operations of the entire silo.

---

Besides connectivity issues, permissions, why the drop index may generate an exception? I can only think of "there wasn't an index already by that name" which is handled appropriately.

Likewise with the creation of an index -- if it cannot create the index for the collection, the stability of reminder collection / infrastructure is undefined. 

A silo failing to start should trigger a rollback on the deployment until the issue resolved.

---

I agree that if the ceration of the index fails, it should fail the reminder service, I already resolved the other comment before your response.
But in this scenario, just because "you can't think of any reason" doesn't mean there are no other reasons. I would rather err on the side of safety. Please swallow the error without re-throwing.